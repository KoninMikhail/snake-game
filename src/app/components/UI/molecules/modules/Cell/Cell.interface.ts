export interface ICellProps {
    index: number;
}
