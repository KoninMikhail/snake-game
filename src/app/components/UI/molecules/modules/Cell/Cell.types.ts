export type CellType = '' | 'fruit' | 'snake' | 'snakeHead';
