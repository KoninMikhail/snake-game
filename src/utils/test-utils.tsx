import React, {PropsWithChildren} from 'react';
import type {RenderOptions} from '@testing-library/react';
import {render} from '@testing-library/react';
import type {PreloadedState} from '@reduxjs/toolkit';
import {configureStore} from '@reduxjs/toolkit';
import {Provider} from 'react-redux';

import type {AppStore, RootState} from '@store/store';
import appSlice from '@store/slices/app.slice';
import soundSlice from '@store/slices/sound.slice';
import maxScoreSlice from '@store/slices/maxScore.slice';
import gameDataSlice from '@store/slices/game.slice';
import {ImagesData} from '@data/imagesData';
import {mockImagesData, mockSoundData} from '@/setupTests';
import SoundsProvider from '@context/SoundsContext';
import ImagesProvider from '@context/ImagesContext';
import {SoundsData} from '@data/soundsData';
import {DefaultTheme, ThemeProvider} from 'styled-components';
import {theme as appTheme} from '@styles/theme';

// As a basic setup, import your same slice reducers

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
    preloadedState?: PreloadedState<RootState>;
    store?: AppStore;
    images?: ImagesData;
    sounds?: SoundsData;
    theme?: DefaultTheme;
}

export function renderWithProviders(
    ui: React.ReactElement,
    {
        preloadedState = {},
        // Automatically create a store instance if no store was passed in
        store = configureStore({
            reducer: {
                app: appSlice,
                sounds: soundSlice,
                maxScore: maxScoreSlice,
                game: gameDataSlice,
            },
            preloadedState,
        }),
        images = mockImagesData,
        sounds = mockSoundData,
        theme = appTheme,
        ...renderOptions
    }: ExtendedRenderOptions = {}
) {
    function Wrapper({children}: PropsWithChildren): JSX.Element {
        return (
            <Provider store={store}>
                <ThemeProvider theme={theme}>
                    <SoundsProvider sounds={sounds}>
                        <ImagesProvider images={images}>{children}</ImagesProvider>
                    </SoundsProvider>
                </ThemeProvider>
            </Provider>
        );
    }

    // Return an object with the store and all of RTL's query functions
    return {store, ...render(ui, {wrapper: Wrapper, ...renderOptions})};
}
