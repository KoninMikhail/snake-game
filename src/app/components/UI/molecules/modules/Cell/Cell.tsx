import {StyledCell, StyledCellContent} from '@ui/molecules/modules/Cell/style/StyledCell';
import {ICellProps} from '@ui/molecules/modules/Cell/Cell.interface';
import useAppSelector from '@hooks/useAppSelector/useAppSelector';
import {gameFruitPositionSelector, gameSnakeHeadSelector, gameSnakePointsSelector,} from '@store/slices/game.slice';

/**
 * @name Cell
 * @author Mikhail Konin <dev.konin@gmail.com>
 *
 * Use for creating gameBoard
 *
 * @param index - current cell index
 * @return {JSX.Element}
 */
export const Cell = ({ index }: ICellProps) => {
    const snakePoints = useAppSelector(gameSnakePointsSelector);
    const snakeHead = useAppSelector(gameSnakeHeadSelector);
    const fruitPos = useAppSelector(gameFruitPositionSelector);

    const isSnake = () => snakePoints.some((snakePoint: number) => snakePoint === index);
    const isSnakeHead = () => snakeHead === index;
    const isFruit = () => fruitPos === index;

    const getCellType = () => {
        if (isSnake()) return 'snake';
        if (isSnakeHead()) return 'snakeHead';
        if (isFruit()) return 'fruit';
    };

    return (
        <StyledCell>
            <StyledCellContent type={getCellType()} />
        </StyledCell>
    );
};
