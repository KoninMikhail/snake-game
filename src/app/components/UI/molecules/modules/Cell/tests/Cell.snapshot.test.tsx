import { renderWithProviders } from '@utils/test-utils';
import { Cell } from '@ui/molecules/modules/Cell/Cell';
import { GameData } from '@store/slices/game.slice';

describe('Cell', () => {
    it('should be equal snapshot with stock cell', () => {
        const { container } = renderWithProviders(<Cell index={1} />);
        expect(container).toMatchSnapshot();
    });
    it('should be equal snapshot with fruit cell', () => {
        const preloadedState = {
            game: { fruit: 15, snake: { points: [12, 13, 14], head: 14 } } as GameData,
        };
        const { container } = renderWithProviders(<Cell index={15} />, { preloadedState });
        expect(container).toMatchSnapshot();
    });

    it('should be equal snapshot with snakeHead cell', () => {
        const preloadedState = {
            game: { fruit: 14, snake: { points: [12, 13, 14], head: 14 } } as GameData,
        };
        const { container } = renderWithProviders(<Cell index={15} />, { preloadedState });
        expect(container).toMatchSnapshot();
    });

    it('should be equal snapshot with snakeSection cell', () => {
        const preloadedState = {
            game: { fruit: 13, snake: { points: [12, 13, 14], head: 14 } } as GameData,
        };
        const { container } = renderWithProviders(<Cell index={15} />, { preloadedState });
        expect(container).toMatchSnapshot();
    });
});
