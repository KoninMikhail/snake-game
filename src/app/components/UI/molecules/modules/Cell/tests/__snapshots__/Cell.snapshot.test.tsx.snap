// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cell should be equal snapshot with fruit cell 1`] = `
.c0 {
  display: grid;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  -webkit-box-pack: stretch;
  -webkit-justify-content: stretch;
  -ms-flex-pack: stretch;
  justify-content: stretch;
  width: 100%;
  height: 100%;
  padding: 0.15em;
}

.c1 {
  display: block;
  width: 100%;
  height: 100%;
  background: black;
  background: white;
  -webkit-animation-name: crdVsA;
  animation-name: crdVsA;
  -webkit-animation-duration: 0.6s;
  animation-duration: 0.6s;
  -webkit-animation-timing-function: ease-out;
  animation-timing-function: ease-out;
  -webkit-animation-direction: alternate;
  animation-direction: alternate;
  -webkit-animation-iteration-count: infinite;
  animation-iteration-count: infinite;
  -webkit-animation-play-state: running;
  animation-play-state: running;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
      type="fruit"
    />
  </div>
</div>
`;

exports[`Cell should be equal snapshot with snakeHead cell 1`] = `
.c0 {
  display: grid;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  -webkit-box-pack: stretch;
  -webkit-justify-content: stretch;
  -ms-flex-pack: stretch;
  justify-content: stretch;
  width: 100%;
  height: 100%;
  padding: 0.15em;
}

.c1 {
  display: block;
  width: 100%;
  height: 100%;
  background: black;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    />
  </div>
</div>
`;

exports[`Cell should be equal snapshot with snakeSection cell 1`] = `
.c0 {
  display: grid;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  -webkit-box-pack: stretch;
  -webkit-justify-content: stretch;
  -ms-flex-pack: stretch;
  justify-content: stretch;
  width: 100%;
  height: 100%;
  padding: 0.15em;
}

.c1 {
  display: block;
  width: 100%;
  height: 100%;
  background: black;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    />
  </div>
</div>
`;

exports[`Cell should be equal snapshot with stock cell 1`] = `
.c0 {
  display: grid;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  -webkit-box-pack: stretch;
  -webkit-justify-content: stretch;
  -ms-flex-pack: stretch;
  justify-content: stretch;
  width: 100%;
  height: 100%;
  padding: 0.15em;
}

.c1 {
  display: block;
  width: 100%;
  height: 100%;
  background: black;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    />
  </div>
</div>
`;
